Mutation docs (#1143)
open-policy-agent/gatekeeper: website/docs/mutation.md

---
id: mutation

> title: Gatekeeper mutation

Reviewer: ```suggestion
title: Mutation
```

---

> The mutation feature allows gatekeeper to not only validate created kubernetes resources, but also modify them based on defined mutation policies.

Reviewer: ```suggestion
The mutation feature allows Gatekeeper to not only validate created Kubernetes resources but also modify them based on defined mutation policies.
```

> The feature is still in an alpha stage, so the final form can still changes. 

Reviewer: Can we add a big "status: alpha" (or pre-alpha) at the top too so it's clear

Reviewer 2: also, nit: "changes" -> "change"

> ## Mutation CRD's

Reviewer: ```suggestion
## Mutation CRDs
```

> The mutation policies are defined by means of mutation specific CRD's: 

Reviewer: ```suggestion
The mutation policies are defined by means of mutation specific CRDs: 
```

- AssignMetadata - defines changes to the metadata section of a resource
- Assign - any change outside the metadata section

> The rules of mutating the metdata section are more strict then for mutating the rest of the resource definition. The differences will be described in more detail in consecutive paragraphs.

Reviewer: ```suggestion
The rules of mutating the metadata section are more strict than for mutating the rest of the resource definition. The differences will be described in more detail in consecutive paragraphs.
```

Reviewer 2: additional nit: "consecutive paragraphs" -> "below"

    match:
        kinds: 
        - APIGroups: []
          kinds: []
        labelSelector: []
        namespaces: []
        namespaceSelector: []
        excludedNamespaces: []
```

Note that the `applyTo` section applies to the Assign CRD only. It allows to filter resoruces by the resource GVK (group version kind).

The `match` section is common to both Assign and AssignMetadata. It supports the following elements:
- kinds - the resource kind, any of the elements listed
- labelSelector - filters resources by resource labels listed
- namespaces - list of allowed namespaces, only resources in listed namespaces will be mutated
- namespaceSelector - filters resources by namespace selector
- excludedNamespaces - list of excluded namespaces, resources in listed namespaces will not be mutated

Note that the resource is not filtered if an element is not present or an empty list.

#### Intent

This specifies what should be changed in the resource.

An example of the section is shown below:
```yaml
    location: "spec.containers[name:foo].imagePullPolicy"
    parameters:
       assign:
        value: "Always"
```

The `location` element specifies the path to be modified.

> The `parametes.assign.valule` element specifies the value to be set for the element specified in `location`. Note that the value can either be a simple string or a composite value.

Reviewer: ```suggestion
The `parameters.assign.value` element specifies the value to be set for the element specified in `location`. Note that the value can either be a simple string or a composite value.
```

 - *spec.containers[name:foo].**imagePullPolicy*** - in the element from the list chosen in the previous step the element `imagePullPolicy` is chosen

The yaml illustrating the above `location`:
```yaml
spec:
   containers:
   -  name:foo
      imagePullPolicy:
```
Note that when the list element is empty (`spec.containers[].imagePullPolicy`), all the list element will be updated.

##### Conditionals

The conditions for updating the resource.
Two types of conditions exist:
- path tests - a resource will only be updated when a specified path exists or not
- value tests - a resource will only be updated when the existing value is/is not contained in a list of values

An example of the conditionals: 
```yaml
parameters:
     pathTests:
          - subPath: "spec.containers[name:foo]"
            condition: MustExist
          - subPath: spec.containers[name:foo].securityContext.capabilities
            condition: MustNotExist

      ifIn: [<lists of current values that cause the value to be modified>]
      ifNotIn: [<lists of current values which will cause the value NOT to be modified>]

```

### AssignMetadata

> The AssignMetadata is a CRD for modifying the metadata section of a resource. Note that the metadata of a resource is a very sensitive piece of data, and inconsiderate updated could result in obnoxious consequences. An example of this could be changing the name or namespace of a resource. The AssignMetadata changes have therefore been limited to only the labels and annotations. Furthermore, it is currently only allowed to add a label or annotation, but not modify an existing one (this might change in leter versions of the feature).

Reviewer: ```suggestion
The AssignMetadata is a CRD for modifying the metadata section of a resource. Note that the metadata of a resource is a very sensitive piece of data, and inconsiderate updates could result in obnoxious consequences. An example of this could be changing the name or namespace of a resource. The AssignMetadata changes have therefore been limited to only the labels and annotations. Furthermore, it is currently only allowed to add a label or annotation, but not modify an existing one (this might change in later versions of the feature).
```

Reviewer 2: additional nits:

"inconsiderate updates" -> "certain mutations"

"obnoxious consequences" -> "unintended consequences"

 
 An example of an AssignMetadata adding a label `owner` set to `admin`:
```yaml
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: AssignMetadata
metadata:
  name: demo-annotation-owner 
spec:
    match:
      scope: Namespaced
    location: "metadata.labels.owner"
    parameters: 
      assign: 

>         value:  "admin"

Reviewer: ```suggestion
        value: "admin"
```

```

## Examples

### Adding an annotation

```yaml
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: AssignMetadata
metadata:
  name: demo-annotation-owner 
spec:
    match:
      scope: Namespaced
    location: "metadata.annotations.owner"
    parameters: 
      assign: 

>         value:  "admin"

Reviewer: ```suggestion
        value: "admin"
```

Below is an example of a simple AssignMetadata CRD:
```yaml
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: AssignMetadata
metadata:
  name: demo-annotation-owner 
spec:
    match:
      scope: Namespaced
      kinds:
        - apiGroups: ["*"]
          kinds: ["Pod"]
    location: "metadata.annotations.owner"
    parameters: 
      assign: 
        value:  "admin"
```

Each mutation CRD can be divided into 3 distinct sections:
- scope - what is to be modified (kinds, namespaces, ...)
- intent - the path an value of the modification
- conditional - conditions under which the mutation will be applied

#### Scope

The scope section describes the resource which will be mutated.
It allows to filter the resources to be mutated by kind, label and namespace.

An example of the scope section.
```yaml
    applyTo:
        - groups: [""]
          kinds: ["Pod"]
          versions: ["v1"]
    match:
        kinds: 
        - APIGroups: []
          kinds: []
        labelSelector: []
        namespaces: []
        namespaceSelector: []
        excludedNamespaces: []
```

> Note that the `applyTo` section applies to the Assign CRD only. It allows to filter resoruces by the resource GVK (group version kind).

Author: Do you have any suggestions on how to describe the difference between appllyTo and match (both having kind and group?) in a not confusing way. I don't think we should describe the difference in how it is implemented.

The `match` section is common to both Assign and AssignMetadata. It supports the following elements:

> - kinds - the resource kind, any of the elements listed

Author: Correct, added

- *spec.**containers[name:foo]**.imagePullPolicy* - container subelement of spec. The container element is a list. Out of the list choosen, an element with the `name` element having the value `foo`.
 - *spec.containers[name:foo].**imagePullPolicy*** - in the element from the list chosen in the previous step the element `imagePullPolicy` is chosen

The yaml illustrating the above `location`:
```yaml
spec:
   containers:
   -  name:foo
      imagePullPolicy:
```
Note that when the list element is empty (`spec.containers[].imagePullPolicy`), all the list element will be updated.

##### Conditionals

The conditions for updating the resource.
Two types of conditions exist:
- path tests - a resource will only be updated when a specified path exists or not
- value tests - a resource will only be updated when the existing value is/is not contained in a list of values

An example of the conditionals: 
```yaml
parameters:
     pathTests:
          - subPath: "spec.containers[name:foo]"
            condition: MustExist
          - subPath: spec.containers[name:foo].securityContext.capabilities
            condition: MustNotExist

>       ifIn: [<lists of current values that cause the value to be modified>]

Reviewer 2: This bit had to change due to the current limits of what we can do with CRDs.

Here is what the current syntax looks like:

```
spec:
   parameters:
      assignIf:
         in: [<value 1>, <value 2>, <value 3>, ...]
         notIn: [<value 1>, <value 2>, <value 3>, ...]
```

      ifNotIn: [<lists of current values which will cause the value NOT to be modified>]

```

### AssignMetadata

> The AssignMetadata is a CRD for modifying the metadata section of a resource. Note that the metadata of a resource is a very sensitive piece of data, and certain mutations could result in unintended consequences. An example of this could be changing the name or namespace of a resource. The AssignMetadata changes have therefore been limited to only the labels and annotations. Furthermore, it is currently only allowed to add a label or annotation, but not modify an existing one (this might change in later versions of the feature).

Reviewer 2: nit: remove "The" from the start of the first sentence.

> The AssignMetadata is a CRD for modifying the metadata section of a resource. Note that the metadata of a resource is a very sensitive piece of data, and certain mutations could result in unintended consequences. An example of this could be changing the name or namespace of a resource. The AssignMetadata changes have therefore been limited to only the labels and annotations. Furthermore, it is currently only allowed to add a label or annotation, but not modify an existing one (this might change in later versions of the feature).

Reviewer 2: Do we want to highlight that this might change? I don't know if we have a plausible path forward for this yet, though discussion is active and ongoing.

We don't need a fully-realized design, but we should have something that people agree is more likely to work than not.

Author: Removed

---
id: mutation
title: Mutation

> status: alpha

Reviewer: I don't think `status` is supported as a header item https://v2.docusaurus.io/docs/markdown-features#markdown-headers

can we move this down?

Author: Moved down to after the first paragraph. Although it's now a bit redundant with what we have in the text.

Here is an example of a simple AssignMetadata CRD:
```yaml
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: AssignMetadata
metadata:
  name: demo-annotation-owner 
spec:
    match:
      scope: Namespaced
      kinds:
        - apiGroups: ["*"]
          kinds: ["Pod"]
    location: "metadata.annotations.owner"
    parameters: 
      assign: 
        value:  "admin"
```

Each mutation CRD can be divided into 3 distinct sections:
- scope - what is to be modified (kinds, namespaces, ...)
- intent - the path and value of the modification
- conditional - conditions under which the mutation will be applied

#### Scope

The scope section describes the resource which will be mutated.
It allows to filter the resources to be mutated by kind, label and namespace.

An example of the scope section.
```yaml
    applyTo:
        - groups: [""]
          kinds: ["Pod"]
          versions: ["v1"]
    match:
        scope: Namespaced | Cluster
        kinds: 
        - APIGroups: []
          kinds: []
        labelSelector: []
        namespaces: []
        namespaceSelector: []
        excludedNamespaces: []
```

> Note that the `applyTo` section applies to the Assign CRD only. It allows filtering of resoruces by the resource GVK (group version kind). Note that the `applyTo` section does not accept globs.

Reviewer: ```suggestion
Note that the `applyTo` section applies to the Assign CRD only. It allows filtering of resources by the resource GVK (group version kind). Note that the `applyTo` section does not accept globs.
```

Author: Done

---
id: mutation
title: Mutation
---

#Gatekeeper mutation

The mutation feature allows Gatekeeper to not only validate created Kubernetes resources but also modify them based on defined mutation policies.
The feature is still in an alpha stage, so the final form can still change.

Status: alpha

## Mutation CRDs

The mutation policies are defined by means of mutation specific CRDs:
- AssignMetadata - defines changes to the metadata section of a resource
- Assign - any change outside the metadata section

The rules of mutating the metadata section are more strict than for mutating the rest of the resource definition. The differences will be described in more detail below.

Here is an example of a simple AssignMetadata CRD:
```yaml
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: AssignMetadata
metadata:
  name: demo-annotation-owner 
spec:
    match:
      scope: Namespaced
      kinds:
        - apiGroups: ["*"]
          kinds: ["Pod"]
    location: "metadata.annotations.owner"
    parameters: 
      assign: 
        value:  "admin"
```

Each mutation CRD can be divided into 3 distinct sections:

> - scope - what is to be modified (kinds, namespaces, ...)

Reviewer 2: nit: this name collides with the `scope: Namespaced` match parameter above.

"selector" maybe? "binding"?

Author: I went with: "Extent of changes"
Selector - could be confused with the namespace selector
binding brings role binding to mind

        value: "Always"
```

The `location` element specifies the path to be modified.
The `parameters.assign.value` element specifies the value to be set for the element specified in `location`. Note that the value can either be a simple string or a composite value.

An example of a composite value:
```yaml
    location: "spec.containers[name:networking]"
    parameters: 
      assign: 
        value: 
          name: "networking"
          imagePullPolicy: Always

```

The `location` element can specify either a simple subelement or an element in a list.
For example the location `spec.containers[name:foo].imagePullPolicy` would be parsed as follows:
- ***spec**.containers[name:foo].imagePullPolicy* - the spec element
- *spec.**containers[name:foo]**.imagePullPolicy* - container subelement of spec. The container element is a list. Out of the list choosen, an element with the `name` element having the value `foo`.
 - *spec.containers[name:foo].**imagePullPolicy*** - in the element from the list chosen in the previous step the element `imagePullPolicy` is chosen

The yaml illustrating the above `location`:
```yaml
spec:
   containers:
   -  name:foo
      imagePullPolicy:
```

> Note that when the list element is empty (`spec.containers[].imagePullPolicy`), all the list element will be updated.

Reviewer 2: I don't think this is true? I think the key field must be provided:

`spec.containers[name:*].imagePullPolicy`

if not, we should fix that

Author: Replaced with:
Wildcards can be used for list element values: `spec.containers[name:*].imagePullPolicy`
I did not check this, but I thought that this is the desired behavior. Let me check that later.

Reviewer 2: We need the key field to make sure users aren't overwriting key values.

I wonder if we are denying requests like:

```
spec.containers[name:*] = {name: "foo", ...}
```

if not, we should be, as wildcards only really make sense when we're setting fields of list members.

---
id: mutation
title: Mutation
---

#Gatekeeper mutation

The mutation feature allows Gatekeeper to not only validate created Kubernetes resources but also modify them based on defined mutation policies.
The feature is still in an alpha stage, so the final form can still change.

Status: alpha

## Mutation CRDs

The mutation policies are defined by means of mutation specific CRDs:
- AssignMetadata - defines changes to the metadata section of a resource
- Assign - any change outside the metadata section

The rules of mutating the metadata section are more strict than for mutating the rest of the resource definition. The differences will be described in more detail below.

Here is an example of a simple AssignMetadata CRD:
```yaml
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: AssignMetadata
metadata:
  name: demo-annotation-owner 
spec:

>     match:

Reviewer 3: ```suggestion
  match:
    scope: Namespaced
    kinds:
      - apiGroups: ["*"]
        kinds: ["Pod"]
    location: "metadata.annotations.owner"
    parameters: 
      assign: 
        value:  "admin"
```

      scope: Namespaced
      kinds:
        - apiGroups: ["*"]
          kinds: ["Pod"]
    location: "metadata.annotations.owner"
    parameters: 
      assign: 
        value:  "admin"
```

Each mutation CRD can be divided into 3 distinct sections:
- extent of changes - what is to be modified (kinds, namespaces, ...)
- intent - the path and value of the modification
- conditional - conditions under which the mutation will be applied

#### Extent of changes

The extent of changes section describes the resource which will be mutated.
It allows to filter the resources to be mutated by kind, label and namespace.

An example of the extent of changes section.
```yaml
    applyTo:

>         - groups: [""]

Reviewer 3: nit: spaces are incorrect

```suggestion
    - groups: [""]
      kinds: ["Pod"]
      versions: ["v1"]
    match:
      scope: Namespaced | Cluster
      kinds: 
      - APIGroups: []
        kinds: []
      labelSelector: []
      namespaces: []
      namespaceSelector: []
      excludedNamespaces: []
```

        kinds: 
        - APIGroups: []
          kinds: []
        labelSelector: []
        namespaces: []
        namespaceSelector: []
        excludedNamespaces: []
```

Note that the `applyTo` section applies to the Assign CRD only. It allows filtering of resources by the resource GVK (group version kind). Note that the `applyTo` section does not accept globs.

The `match` section is common to both Assign and AssignMetadata. It supports the following elements:
- scope - the scope (Namespaced | Cluster) of the mutated resource
- kinds - the resource kind, any of the elements listed
- labelSelector - filters resources by resource labels listed
- namespaces - list of allowed namespaces, only resources in listed namespaces will be mutated
- namespaceSelector - filters resources by namespace selector
- excludedNamespaces - list of excluded namespaces, resources in listed namespaces will not be mutated

Note that the resource is not filtered if an element is not present or an empty list.

#### Intent

This specifies what should be changed in the resource.

An example of the section is shown below:
```yaml
    location: "spec.containers[name:foo].imagePullPolicy"
    parameters:

>        assign:

Reviewer 3: ```suggestion
      assign:
        value: "Always"
```

        value: "Always"
```

The `location` element specifies the path to be modified.
The `parameters.assign.value` element specifies the value to be set for the element specified in `location`. Note that the value can either be a simple string or a composite value.

An example of a composite value:
```yaml
    location: "spec.containers[name:networking]"
    parameters: 
      assign: 
        value: 
          name: "networking"
          imagePullPolicy: Always

```

The `location` element can specify either a simple subelement or an element in a list.
For example the location `spec.containers[name:foo].imagePullPolicy` would be parsed as follows:
- ***spec**.containers[name:foo].imagePullPolicy* - the spec element
- *spec.**containers[name:foo]**.imagePullPolicy* - container subelement of spec. The container element is a list. Out of the list choosen, an element with the `name` element having the value `foo`.
 - *spec.containers[name:foo].**imagePullPolicy*** - in the element from the list chosen in the previous step the element `imagePullPolicy` is chosen

The yaml illustrating the above `location`:
```yaml
spec:

>    containers:

Reviewer 3: ```suggestion
  containers:
  - name:foo
    imagePullPolicy:
```

   -  name:foo
      imagePullPolicy:
```

Wildcards can be used for list element values: `spec.containers[name:*].imagePullPolicy`

##### Conditionals

The conditions for updating the resource.
Two types of conditions exist:
- path tests - a resource will only be updated when a specified path exists or not
- value tests - a resource will only be updated when the existing value is/is not contained in a list of values

An example of the conditionals: 
```yaml
parameters:

>      pathTests:

Reviewer 3: ```suggestion
  pathTests:
    - subPath: "spec.containers[name:foo]"
        condition: MustExist
    - subPath: spec.containers[name:foo].securityContext.capabilities
        condition: MustNotExist
  assignIf:
    in: [<value 1>, <value 2>, <value 3>, ...]
    notIn: [<value 1>, <value 2>, <value 3>, ...]
```

          - subPath: "spec.containers[name:foo]"
            condition: MustExist
          - subPath: spec.containers[name:foo].securityContext.capabilities
            condition: MustNotExist

     assignIf:
         in: [<value 1>, <value 2>, <value 3>, ...]
         notIn: [<value 1>, <value 2>, <value 3>, ...]

```

### AssignMetadata

AssignMetadata is a CRD for modifying the metadata section of a resource. Note that the metadata of a resource is a very sensitive piece of data, and certain mutations could result in unintended consequences. An example of this could be changing the name or namespace of a resource. The AssignMetadata changes have therefore been limited to only the labels and annotations. Furthermore, it is currently only allowed to add a label or annotation.
 
 An example of an AssignMetadata adding a label `owner` set to `admin`:
```yaml
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: AssignMetadata
metadata:
  name: demo-annotation-owner 
spec:

>     match:

Reviewer 3: ```suggestion
  match:
    scope: Namespaced
  location: "metadata.labels.owner"
  parameters: 
    assign: 
      value: "admin"
```

      scope: Namespaced
    location: "metadata.labels.owner"
    parameters: 
      assign: 
        value: "admin"
```

## Examples

### Adding an annotation

```yaml
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: AssignMetadata
metadata:
  name: demo-annotation-owner 
spec:

>     match:

Reviewer 3: ```suggestion
  match:
    scope: Namespaced
  location: "metadata.annotations.owner"
  parameters: 
    assign: 
      value: "admin"
```

      scope: Namespaced
    location: "metadata.annotations.owner"
    parameters: 
      assign: 
        value: "admin"
```

### Setting security context of a specific container in a Pod in a namespace to be privileged

Set the security context of container named `foo` in a Pod in namespace `bar` to be privileged

```yaml
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: Assign
metadata:
  name: demo-privileged 
  namespace: default
spec:

>     match:

Reviewer 3: ```suggestion
  match:
    scope: Namespaced
    kinds:
      - apiGroups: ["*"]
        kinds: ["Pod"]
    namespaces: ["bar"]
  location: "spec.containers[name:foo].securityContext.privileged"
  parameters: 
    assign: 
      value: false
```

      scope: Namespaced
      kinds:
        - apiGroups: ["*"]
          kinds: ["Pod"]
      namespaces: ["bar"]
    location: "spec.containers[name:foo].securityContext.privileged"
    parameters: 
      assign: 
        value: false
```

#### Setting imagePullPolicy of all containers to Always in all namespaces except namespace `system`

```yaml
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: Assign
metadata:
  name: demo-image-pull-policy
  namespace: default
spec:

>     match:

Reviewer 3: ```suggestion
  match:
    scope: Namespaced
    kinds:
      - apiGroups: ["*"]
        kinds: ["Pod"]
    excludedNamespaces: ["system"]
  location: "spec.containers[name:*].imagePullPolicy"
  parameters: 
    assign: 
      value: Always
```

      scope: Namespaced
      kinds:
        - apiGroups: ["*"]
          kinds: ["Pod"]
      excludedNamespaces: ["system"]
    location: "spec.containers[name:*].imagePullPolicy"
    parameters: 
      assign: 
        value: Always
```

### Adding a `network` sidecar to a Pod

```yaml
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: Assign
metadata:
  name: demo-sidecar
  namespace: default
spec:

>     match:

Reviewer 3: ```suggestion
  match:
    scope: Namespaced
    kinds:
      - apiGroups: ["*"]
        kinds: ["Pod"]
  location: "spec.containers[name:networking]"
  parameters: 
    assign: 
      value: 
        name: "networking"
        imagePullPolicy: Always
        image: quay.io/foo/bar:latest
        command: ["/bin/bash", "-c", "sleep INF"]

```

      scope: Namespaced
      kinds:
        - apiGroups: ["*"]
          kinds: ["Pod"]
    location: "spec.containers[name:networking]"
    parameters: 
      assign: 
        value: 
          name: "networking"
          imagePullPolicy: Always
          image: quay.io/foo/bar:latest
          command: ["/bin/bash", "-c", "sleep INF"]

```

### Adding dnsPolicy and dnsConfig to a Pod

```yaml
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: Assign
metadata:
  name: demo-dns-policy
  namespace: default
spec:

>     match:

Reviewer 3: ```suggestion
  match:
    scope: Namespaced
    kinds:
      - apiGroups: ["*"]
        kinds: ["Pod"]
  location: "spec.dnsPolicy"
  parameters: 
    assign: 
      value: None
```

      scope: Namespaced
      kinds:
        - apiGroups: ["*"]
          kinds: ["Pod"]
    location: "spec.dnsPolicy"
    parameters: 
      assign: 
        value: None
---
apiVersion: mutations.gatekeeper.sh/v1alpha1
kind: Assign
metadata:
  name: demo-dns-config
  namespace: default
spec:

>     match:

Reviewer 3: ```suggestion
  match:
    scope: Namespaced
    kinds:
      - apiGroups: ["*"]
        kinds: ["Pod"]
  location: "spec.dnsConfig"
  parameters:
    assign:
      value:
        nameservers: 
        - 1.2.3.4 
```

Author: Fixed indentation (for all the comments)
